Skip procedures while tokenizing if not used under contains

parth121101 opened the pull request:
Problem: 
```
PROGRAM continue_compilation_ff_2
print *, "This is a test program"
SUBROUTINE faulty_subroutine(a, b, c)
INTEGER, INTENT(IN) :: sub_a
END SUBROUTINE faulty_subroutine
END PROGRAM continue_compilation_ff_2
```

The subroutine here is not defined under the contains keyword. Therefore, the entire subroutine definition is incorrect, and we need to skip this procedure.

To handle this, I added logic to skip all lines of a procedure's definition if it is not placed under the contains section, continuing until we encounter the corresponding end statement of the program.

parth121101 commented on tests/reference/asr-continue_compilation_ff_2-08ef965.json:
```diff
@@ -0,0 +1,13 @@
+    "stderr": null,
+    "stderr_hash": null,
+    "returncode": 0
+}
```
will check this.

parth121101 commented:
Problem: 
Procedure declared inside main program without the use of `contains` keyword. 

Proposed solution:
1) Identify that a procedure (subroutine or function) is defined inside main program and `contains` keyword is missed.
2) Skip all the statements from the first line where any procedure definition is first encountered till the end program is found.
3) If continue compilation is ON then skip all this statements and add an error inside diagnostics. 
4) If continue compilation is OFF then just throw an error. 


What is left here?
Currently in the latest commit I used --show-asr in the tests options. Initially I was using --show-ast which was unable to return non zero exit code tough diagnostics has some errors. Will find the bug for this and figure out a fix.

@Pranavchiku @certik Can I get a review please. Let me know if this is a correct implementation or there's some flaw here.

Pranavchiku commented:
Marked it as draft.

parth121101 commented on run_tests.py:
```diff
@@ -432,6 +433,16 @@ def is_included(backend):
+                filename,
+                update_reference,
+                verify_hash,
+                extra_args)
```
This is resolved.

Pranavchiku commented on src/lfortran/parser/fixedform_tokenizer.cpp:
```diff
@@ -1584,6 +1584,24 @@ struct FixedFormRecursiveDescent {
             next_line(cur); // Does not generate any code?
             while(lex_procedure(cur));
+        } else if (next_is(cur, "subroutine") || next_is(cur, "function")) {
+            Location loc;
+            loc.first = cur - string_start;
+            loc.last = cur - string_start;
```
can we not use `token_loc(loc)`?

parth121101 replied:
`token_loc` function has following definition:
```
loc.first = tok-string_start;
loc.last = cur-string_start-1;
```


While the one used here is a bit different. If I try to use the function `token_loc` it starts to fail. So what should I consider here 
-->> A bug in `token_loc` function definition 
-->> Or a bug in my implementation of storing location information


```
lfortran /Users/parth1211/repos/lfortran/tests/errors/continue_compilation_ff_2.f --continue-compilation --show-ast --fixed-form
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  File "/Users/parth1211/repos/lfortran/src/bin/lfortran.cpp", line 2720
    LCompilers::print_stack_on_segfault();
  File "/Users/parth1211/repos/lfortran/src/bin/lfortran.cpp", line 2519
    return emit_ast(opts.arg_file, compiler_options);
  File "/Users/parth1211/repos/lfortran/src/bin/lfortran.cpp", line 586
    std::cerr << diagnostics.render(lm, compiler_options);
  File "/Users/parth1211/repos/lfortran/src/libasr/diagnostics.cpp", line 86
    out += render_diagnostic_human(d, lm, compiler_options.use_colors,
  File "/Users/parth1211/repos/lfortran/src/libasr/diagnostics.cpp", line 165
    populate_spans(d, lm);
  File "/Users/parth1211/repos/lfortran/src/libasr/diagnostics.cpp", line 151
    for (auto &s : l.spans) {
  File "/Users/parth1211/repos/lfortran/src/libasr/diagnostics.cpp", line 145
    LCOMPILERS_ASSERT(s.source_code.size() > 0)
AssertFailed: s.source_code.size() > 0
```

parth121101 replied:
Issue link: https://github.com/lfortran/lfortran/issues/7103

Pranavchiku reviewed (approved):
left a few comments, rest all looks good

parth121101 commented on run_tests.py:
```diff
@@ -432,6 +434,26 @@ def is_included(backend):
+                filename,
+                update_reference,
+                verify_hash,
+                extra_args)
```
I wanted to print ast as well for the added test. Hence I created a new flag for the same. Do you want me to add `fixed_form_cc_ast` flag for all the tests and remove `fixed_form_cc` ?

parth121101 replied:
Surely will address this changes. 
Yes this was not working before. As discussed I was supposed to fix this problem in a separate PR and rebase this PR on top. PR for the same: https://github.com/lfortran/lfortran/pull/7089

parth121101 replied:
A separate PR to return AST and upgrading test refs. PR: https://github.com/lfortran/lfortran/pull/7106

parth121101 commented on run_tests.py:
```diff
@@ -432,6 +434,26 @@ def is_included(backend):
                 verify_hash,
                 extra_args)
+        
+    if fixed_form_cc_asr:
```
Will remove this flag

parth121101 commented:
1) Removed the flag `fixed_form_cc_ast` from this PR. Will create a separate PR where I will use `fixed_form_cc_ast` flag for all the tests and discard `fixed_form_cc` flag.

2) Opened a new bug where `token_loc` needs to be debugged and fixed. Issue link: https://github.com/lfortran/lfortran/issues/7103

Pranavchiku reviewed (approved):
We have all TODOs and issues opened, I'm fine to merge this and improve in subsequent PRs.
